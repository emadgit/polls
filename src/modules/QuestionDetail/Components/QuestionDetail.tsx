import React, { useEffect, FunctionComponent } from "react";
import { Question } from "../../../Interface/Question";
import { Choice } from "../../../Interface/Choice";
import Paper from "@material-ui/core/Paper";
import Typography from "@material-ui/core/Typography";
import Grid from "@material-ui/core/Grid";
import Button from "@material-ui/core/Button";
import Chip from "@material-ui/core/Chip";
import { WrappedProps } from "../Containers/QuestionDetail";
import Snackbar from "@material-ui/core/Snackbar";
import SnackbarContent from "@material-ui/core/SnackbarContent";

export const QuestionDetail: FunctionComponent<WrappedProps> = ({
  handleVote,
  showSnackbar,
  handleClose,
  location
}) => {
  console.log("WTF : ", location.state);
  const data: Question = location.state.question;
  const renderChoices =
    data &&
    data.choices.map((c: Choice) => {
      return (
        <Grid item sm={6} xs={12} md={2} key={c.choice}>
          <Button variant="flat" onClick={() => handleVote(c.url)}>
            <Chip
              color="default"
              label={`
                     ${c.choice} | ${c.votes}

                    `}
            />
          </Button>
        </Grid>
      );
    });
  return (
    <div className="viewSection">
      <Typography variant="title" component="h3" className="whiteColor">
        {data.question}
      </Typography>
      <Grid container spacing={8} className="verticalSpacing">
        {renderChoices}
      </Grid>
      <Snackbar
        anchorOrigin={{
          horizontal: "right",
          vertical: "bottom"
        }}
        autoHideDuration={2000}
        open={showSnackbar}
        onClose={handleClose}
        ContentProps={{
          "aria-describedby": "message-id"
        }}
      >
        <SnackbarContent
          style={{ backgroundColor: "green" }}
          message={<span id="message-id">Your vote counted!</span>}
        />
      </Snackbar>
    </div>
  );
};
