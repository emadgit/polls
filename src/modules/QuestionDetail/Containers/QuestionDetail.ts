import { withHandlers, withState, compose } from "recompose";
import { QuestionDetail as Component } from "../Components/QuestionDetail";
import { RouteComponentProps } from "react-router";
import { config } from "../../../config";
import { postData } from "../../../utils";
import { Question } from '../../../Interface/Question';
import cloneDeep from 'lodash/cloneDeep';

interface Handler {
  handleVote: (cId: string, i: number) => void;
  handleClose: () => void;
}

interface State {
  showSnackbar: boolean;
  setShowSnackbar: (value: boolean) => void;
}

interface QuestionState {
    question: Question,
    setQuestion: (value: Question) => void;
}

const withStateHOC = withState("showSnackbar", "setShowSnackbar", false);
const withQuestionStateHOC = withState("question", "setQuestion", {});

const withHandlerHOC = withHandlers<WrappedProps, Handler>({
  handleVote: ({ setShowSnackbar, setQuestion, question }) => (cId: string, i:number) => {
    postData(`${config.API_URL}${cId}`)
      .then(data => {
        let cloneItem = cloneDeep(question)
        cloneItem.choices[i].votes = data.votes;
        setQuestion(cloneItem)
        setShowSnackbar(true);
      }) // JSON-string from `response.json()` call
      .catch(error => console.error(error));
  },
  handleClose: ({ history, setShowSnackbar }) => () => {
    setShowSnackbar(false);
    history.goBack();
  }
});

export type WrappedProps = RouteComponentProps & Handler & State & QuestionState;

const enhance = compose<WrappedProps, {}>(
  withStateHOC,
  withQuestionStateHOC,
  withHandlerHOC
);

export const QuestionDetail = enhance(Component);
