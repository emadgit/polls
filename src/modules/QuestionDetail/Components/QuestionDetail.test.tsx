import * as React from "react";
import { shallow } from "enzyme";
import { QuestionDetail } from "./QuestionDetail";

let mock: any = jest.fn();

const mockData = {
  history: { mock },
  location: {
    state: {
      question: {
        choices: [
          {
            choice: "TEST",
            url: "/questions/7/choices/54",
            votes: 213
          }
        ],
        published_at: "2015-05-27T21:22:26.601000+00:00",
        question: "Bacon?",
        url: "/questions/7"
      }
    }
  },
  match: { mock },
  handleVote: jest.fn(),
  handleClose: { mock },
  setShowSnackbar: { mock },
  showSnackbar: { mock }
};

// @ts-ignore
const component = shallow(<QuestionDetail {...mockData} />);

describe("QuestionDetail Component", () => {
  it("should render correctly", () => {
    expect(component).toMatchSnapshot();
  });
});
