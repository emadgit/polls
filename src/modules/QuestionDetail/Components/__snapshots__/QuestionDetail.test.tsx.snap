// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionDetail Component should render correctly 1`] = `
<div
  className="viewSection"
>
  <WithStyles(Typography)
    className="whiteColor padding-10"
    component="h3"
    variant="title"
  />
  <WithStyles(Grid)
    className="verticalSpacing"
    container={true}
    spacing={8}
  />
  <WithStyles(Snackbar)
    ContentProps={
      Object {
        "aria-describedby": "message-id",
      }
    }
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "bottom",
      }
    }
    autoHideDuration={2000}
    onClose={
      Object {
        "mock": [MockFunction],
      }
    }
    open={
      Object {
        "mock": [MockFunction],
      }
    }
  >
    <WithStyles(SnackbarContent)
      message={
        <span
          id="message-id"
        >
          Your vote counted!
        </span>
      }
      style={
        Object {
          "backgroundColor": "green",
        }
      }
    />
  </WithStyles(Snackbar)>
</div>
`;
